### ISSUE TITLE
parallel code with prange returns a wrong results with two successive loops

### ISSUE THREAD

**Issue by arnaudlegout** [numba/numba#9829]
I noticed a bug running parallel code with prange when I have two successive loops.
Here is the code to reproduce the issue

```
import numba
import numpy as np

@numba.njit
def correct_test_sequential(index, values, max_id):
    MAX_INT = np.iinfo(np.uint32).max
    array = np.ones(max_id + 1, dtype=np.uint32) * MAX_INT

    for i in range(len(index)):
        array[index[i]] = values[i]

    max_scc = values.max()
    for i in range(len(array)):
        if array[i] == MAX_INT:
            array[i] = max_scc + 1
            max_scc += 1

    return array

@numba.njit(parallel=True)
def buggy_test_parallel(index, values, max_id):
    MAX_INT = np.iinfo(np.uint32).max
    array = np.ones(max_id + 1, dtype=np.uint32) * MAX_INT

    for i in numba.prange(len(index)):
        array[index[i]] = values[i]

    max_scc = values.max()
    for i in numba.prange(len(array)):
        if array[i] == MAX_INT:
            array[i] = max_scc + 1
            max_scc += 1

    return array



print(correct_test_sequential(
    np.array([9, 11, 12, 15, 19]),
    np.array([0, 1, 2, 3, 4]),
    20,
))

print(buggy_test_parallel(
    np.array([9, 11, 12, 15, 19]),
    np.array([0, 1, 2, 3, 4]),
    20,
))
```

and here is the result
```
In [8]: %run -i test.py
[ 5  6  7  8  9 10 11 12 13  0 14  1  2 15 16  3 17 18 19  4 20]
[ 1  2  3  4  5  6  7  8  9  0 10  1  2 11 12  3  1  1  1  4  2]

In [9]: numba.__version__
Out[9]: '0.60.0'
```

Both codes make exactly the same computation (I just replaced numba.prange by range), but the returned array are different. The sequential code gives the correct result.

**Comment by guilhermeleobas** [comment]
Hi @arnaudlegout, thanks for the report.

This is not a bug in Numba itself. There's a race condition in `buggy_test_parallel`, as different threads can read and modify `max_scc`.

**Comment by arnaudlegout** [comment]
What is spawned and how parallel loops work is not well documented. In my case, `max_scc = values.max()` is supposed to be constant and equal to 5, so whatever the race condition, how is it possible that `max_scc` starts at one. I perfectly understand that with threads, I do not have order guarantee, but `max_scc` is only incremented, and `max_scc` is initialized at 5.

**Comment by guilhermeleobas** [comment]
There's a race condition in the accumulator `max_scc`. There's no guarantee a thread will wait for another to finish increment it. So, you may end in a scenario where a thread X will read a value of `max_scc` before another thread Y writes into it.

**Comment by guilhermeleobas** [comment]
I don't know a lot about the parallel target. But given the undefined behavior, Numba is probably generating wrong code as well. If one comments the line `max_scc += 1`, the output is something more reasonable.

cc @toddrme2178

**Comment by guilhermeleobas** [comment]
I don't know a lot about the parallel target. But given the undefined behavior, Numba is probably generating wrong code as well. If one comments the line `max_scc += 1`, the output is something more reasonable.

cc @DrTodd13

**Comment by arnaudlegout** [comment]
Yes, this is my point, and this is why I believe it is a bug. A race condition could be explained by tracking the scheduling of the threads. Here something different happens that is not solely linked to a thread scheduling issue. 

If my code is doing something unspecified for numba, I would expect an exception. But I don't see why numba could not correctly initialize the `max_scc` value.

Also I had a second look at the numba documentation for parallel execution. The documentation gives interesting explanations (in particular for reductions), but I did not find how is working two consecutive parallel loops. My reasonable guess is that as long as all threads for the first loop are not complete, we do not start the second loop.
